Libraries/MLXLMCommon/Evaluate.swift

```diff
@@ -0,0 +1,556 @@
+    }
+
+    private let compiledCategorical: (MLXArray, MLXArray) -> MLXArray = {
+        compile(inputs: [MLXRandom.globalState], outputs: [MLXRandom.globalState]) { logits, temp in
+            categorical(logits * (1 / temp))
+        }
+    }()
+
+    public func sample(logits: MLXArray) -> MLXArray {
+        compiledCategorical(logits, temp)
+    }
+}
+
+/// Processor that implements a `repetitionPenalty`
+public struct RepetitionContext: LogitProcessor {
+    /// tokens in the repetition context sliding window
+    var tokens = [Int]()
+
+    /// current write into into the tokens circular array
+    var index = 0
+
+    /// penalty factor for repeating tokens
+    let repetitionPenalty: Float
+
+    /// number of tokens to consider for repetition penalty
+    let repetitionContextSize: Int
+
+    init(repetitionPenalty: Float, repetitionContextSize: Int) {
+        precondition(repetitionContextSize > 0)
+        self.repetitionPenalty = repetitionPenalty
+        self.repetitionContextSize = repetitionContextSize
+    }
+
+    mutating public func prompt(_ prompt: MLXArray) {
+        if prompt.shape[0] <= repetitionContextSize {
+            self.tokens = prompt.asArray(Int.self)
+        } else {
+            self.tokens = prompt[(-repetitionContextSize)...].asArray(Int.self)
+        }
+    }
+
+    public func process(logits: MLXArray) -> MLXArray {
+        if tokens.count > 0 {
+            let indices = MLXArray(tokens.map { UInt32($0) })
+            var selectedLogits = logits[0..., indices]
+
+            selectedLogits = MLX.where(
+                selectedLogits .< 0, selectedLogits * repetitionPenalty,
+                selectedLogits / repetitionPenalty)
+
+            logits[0..., indices] = selectedLogits
+            return logits
+        }
+
+        return logits
+    }
+
+    mutating public func didSample(token: MLXArray) {
+        if tokens.count >= repetitionContextSize {
+            tokens[index] = token.item(Int.self)
+            index = (index + 1) % repetitionContextSize
+        } else {
+            tokens.append(token.item(Int.self))
+        }
+    }
+}
+
+/// Generator of tokens.
+///
+/// This is typically used via a call to ``generate(input:parameters:context:didGenerate:)``.
+///
+/// To use it directly:
+///
+/// ```swift
+/// let generateParameters: GenerateParameters
+/// let input: LMInput
+/// let model: LanguageModel
+///
+/// let iterator = try TokenIterator(input: input, model: model, parameters: parameters)
+///
+/// for token in iterator {
+///     ...
+/// }
+/// ```
+///
+/// Tokens are integers that can be passed through a `Tokenizer` or ``StreamingDetokenizer`` to produce Strings.
+///
+/// Port of `generate_step()` from https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/utils.py
+///
+/// Note: this uses `asyncEval()` and there may be an async evaluation running after a call to `next()`.
+public struct TokenIterator: Sequence, IteratorProtocol {
```

Comment:
The same `TokenIterator` as before, but refactored to handle optional `State` that VLMs might return.  This now deals with `LMInput` (a struct that encapsulates a prompt and any media) and produces `LMOutput` (a struct that holds minimally a `logits` MLXArray but potentially other values as needed for VLMs).